### PR TITLE
Modifications and improvements to the base functionalities

### PR DESCRIPTION
**Pull request by S-Dafarra**
This is the first of three PRs to merge the modifications done in the ``humanoidPlanning`` branch

Among the notable modifications, we have:
- Added the callback for Opti
- Definition of ``OverridableParameter`` and ``OverridableVariable``. This is to allow defining some set of predefined composite variables, leaving to the user the possibility to decide of the components need to be considered variables or parameters
- Added the possibility to add expressions for the entire horizon in the ``MultipleShootingSolver``. In order to do this, the ``OptimalControlProblem`` outputs also a symbolic structure, where the structure matches the user-defined one, but without being expanded over time.

Note that some modifications to the dependencies will be needed in the following PRs.

### CODE REVIEW

File: .github/workflows/ci_cd.yml [hunk 1]
```diff
@@ -31,11 +31,15 @@ jobs:
...
           miniforge-version: latest
+          channels: conda-forge,robotology
+          channel-priority: true
 
       - name: Dependencies
         shell: bash -l {0}
         run: |
-          mamba install python=${{ matrix.python }} casadi pytest
+          mamba install python=${{ matrix.python }} casadi pytest liecasadi adam-robotics idyntree meshcat-python ffmpeg-python matplotlib
+          mamba install metis=5.1.0 # Reminder for https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/4857#issuecomment-1699470569
```
Comment (traversaro) [inline .github/workflows/ci_cd.yml:41]: I think this is not necessary anymore after https://github.com/conda-forge/conda-forge-repodata-patches-feedstock/pull/527 .

Comment (S-Dafarra) [reply .github/workflows/ci_cd.yml]: Thanks, fixed in https://github.com/ami-iit/hippopt/pull/9/commits/535ce016d1620e06bbc4eef52de66d5de7d9f990

**Review by rob-mau — APPROVED** [review]

**Review by diegoferigo — APPROVED** [review]
LGTM, nice work! I got a bit lost in the logic of using variables names as I never used this functionality firsthand, but I trust your tests on this. I'm glad to see that the pattern using `dataclasses` with custom metadata is eventually working pretty well providing also a good flexibility.

**Comment by S-Dafarra** [comment]
Thank you all, merging and moving to the second 😊

**Pull request merged**